code/controllers/subsystem/movement/movement.dm

```diff
@@ -22,30 +33,71 @@ SUBSYSTEM_DEF(movement)
 	var/our_name = typenames[length(typenames)] //Get the last name in the list, IE the subsystem identifier
 
 	var/datum/controller/subsystem/movement/old_version = global.vars["SS[our_name]"]
-	processing = old_version.processing
-	currentrun = old_version.currentrun
+	buckets = old_version.buckets
+	sorted_buckets = old_version.sorted_buckets
 
 /datum/controller/subsystem/movement/fire(resumed)
 	if(!resumed)
 		canonical_time = world.time
-		currentrun = processing.Copy()
-
-	var/list/running = currentrun //Cache for... you've heard this before
-	while(running.len)
-		var/datum/move_loop/loop = running[running.len]
-		running.len--
-		if(loop.timer <= canonical_time)
-			loop.process() //This shouldn't get nulls, if it does, runtime
-		if (MC_TICK_CHECK)
+
+	for(var/list/bucket_info in sorted_buckets)
+		var/time = bucket_info[MOVEMENT_BUCKET_TIME]
+		if(time > canonical_time || MC_TICK_CHECK)
 			return
+		pour_bucket(bucket_info)
+
+/datum/controller/subsystem/movement/proc/pour_bucket(list/bucket_info)
+	var/list/processing = bucket_info[MOVEMENT_BUCKET_LIST] // Cache for lookup speed
+	while(processing.len)
+		var/datum/move_loop/loop = processing[processing.len]
+		processing.len--
+		loop.process() //This shouldn't get nulls, if it does, runtime
+		if(!QDELETED(loop)) //Re-Insert the loop
+			loop.timer = world.time + loop.delay
+			queue_loop(loop)
+		if (MC_TICK_CHECK)
+			break
+
+	if(length(processing))
+		return // Still work to be done
+	var/bucket_time = bucket_info[MOVEMENT_BUCKET_TIME]
+	remove_bucket(bucket_time)
 	visual_delay = MC_AVERAGE_FAST(visual_delay, max((world.time - canonical_time) / wait, 1))
 
+/datum/controller/subsystem/movement/proc/remove_bucket(bucket_time)
+	for(var/i in 1 to length(sorted_buckets))
+		var/list/bucket_info = sorted_buckets[i]
+		if(bucket_info[MOVEMENT_BUCKET_TIME] != bucket_time)
+			continue
+		sorted_buckets.Cut(i, i + 1) //Removes just this list
+		break
+	//Removes the assoc lookup too
+	buckets -= "[bucket_time]"
+
+/datum/controller/subsystem/movement/proc/queue_loop(datum/move_loop/loop)
+	var/target_time = loop.timer
+	var/string_time = "[target_time]"
+	if(buckets[string_time])
+		buckets[string_time] += loop
+	else
+		buckets[string_time] = list(loop)
+		// This makes buckets and sorted buckets point to the same place, allowing for quicker inserts
+		var/list/new_bucket = list(list(target_time, buckets[string_time]))
+		BINARY_INSERT_DEFINE(new_bucket, sorted_buckets, , list(target_time), FIRST_INDEX, COMPARE_KEY)
```

Comment:
the spacemanDMM is very unhappy with this

Reply:
I have killed spacemaniac